perception/autoware_lidar_frnet/lib/preprocess/preprocess_kernel.cu

```diff
@@ -0,0 +1,239 @@
+    if (
+      proj_idxs[(y * const_interpolation.w) + (x - 1)] &&
+      proj_idxs[(y * const_interpolation.w) + (x + 1)]) {
+      // Interpolate points
+      auto input_data4 = reinterpret_cast<float4 *>(proj_2d);
+      float4 left = input_data4[idx - 1];
+      float4 right = input_data4[idx + 1];
+
+      // Compute the mean of the valid neighbors
+      const auto interpolated_point = make_float4(
+        (left.x + right.x) * 0.5f, (left.y + right.y) * 0.5f, (left.z + right.z) * 0.5f,
+        (left.w + right.w) * 0.5f);
+
+      auto append_idx = atomicAdd(output_num_points, 1);
+      output_points[append_idx * 4 + 0] = interpolated_point.x;
+      output_points[append_idx * 4 + 1] = interpolated_point.y;
+      output_points[append_idx * 4 + 2] = interpolated_point.z;
+      output_points[append_idx * 4 + 3] = interpolated_point.w;
+
+      const auto proj_coor = project2d(
+        make_float3(interpolated_point.x, interpolated_point.y, interpolated_point.z),
+        const_frustum.w, const_frustum.h);
+
+      output_coors[append_idx * 3 + 0] = 0;
+      output_coors[append_idx * 3 + 1] = proj_coor.y;
+      output_coors[append_idx * 3 + 2] = proj_coor.x;
+
+      output_coors_keys[append_idx] = proj_coor.y * const_frustum.w + proj_coor.x;
+    }
+  }
+}
+
+cudaError_t PreprocessCuda::interpolatePoints_launch(
+  uint32_t * proj_idxs, float * proj_2d, uint32_t * output_num_points, float * output_points,
+  int64_t * output_coors, int64_t * output_coors_keys)
+{
+  dim3 block(
+    utils::divup(interpolation_.w, utils::kernel_2d_size),
+    utils::divup(interpolation_.h, utils::kernel_2d_size));
+  dim3 threads(utils::kernel_2d_size, utils::kernel_2d_size);
+
+  interpolatePoints_kernel<<<block, threads, 0, stream_>>>(
+    proj_idxs, proj_2d, output_num_points, output_points, output_coors, output_coors_keys);
+
+  return cudaGetLastError();
+}
+
+void PreprocessCuda::generateUniqueCoors(
+  const uint32_t num_points, const int64_t * coors, const int64_t * coors_keys,
+  uint32_t & output_num_unique_coors, int64_t * output_voxel_coors, int64_t * output_inverse_map)
+{
+  // Input unordered coors
+  thrust::device_vector<Coord> coors_d(
+    reinterpret_cast<const Coord *>(coors), reinterpret_cast<const Coord *>(coors) + num_points);
+
+  // Original indices of coors
+  thrust::device_vector<int64_t> sorted_idxs(num_points);
+  thrust::sequence(thrust::device, sorted_idxs.begin(), sorted_idxs.end(), 0);
+
+  // Sort indices by coors keys
+  thrust::device_vector<int64_t> coors_keys_d(coors_keys, coors_keys + num_points);
+  thrust::stable_sort_by_key(
+    thrust::device, coors_keys_d.begin(), coors_keys_d.end(), sorted_idxs.begin());
+
+  // Reorder coors based on the sorted indices
+  thrust::device_vector<Coord> coors_sorted(num_points);
+  thrust::gather(
+    thrust::device, sorted_idxs.begin(), sorted_idxs.end(), coors_d.begin(), coors_sorted.begin());
+
+  // Generate inverse_map for the sorted coors (cannot be parallelized)
```

Comment:
I am not really sure, but could you take a brief look at @manato 's code for the voxelization, tensorrt's unique plugin, and ptv3's preprocessing? Kind of similar and gives me the impression that could be potentially faster